src/shared/cache-types.cts

```diff
@@ -1,58 +1,133 @@
 } from 'next/dist/server/response-cache/types.js'
 
-export type {
-  CacheHandler,
-  CacheHandlerContext,
-  IncrementalCache,
-} from 'next/dist/server/lib/incremental-cache/index.js'
+export type { CacheHandlerContext } from 'next/dist/server/lib/incremental-cache/index.js'
 
-export type NetlifyCachedRouteValue = Omit<CachedRouteValue, 'body'> & {
+/**
+ * Shape of the cache value that is returned from CacheHandler.get or passed to CacheHandler.set
+ */
+type CachedRouteValueForMultipleVersions = Omit<CachedRouteValue, 'kind'> & {
+  kind: 'ROUTE' | 'APP_ROUTE'
+}
+
+/**
+ * Used for storing in blobs and reading from blobs
+ */
+export type NetlifyCachedRouteValue = Omit<CachedRouteValueForMultipleVersions, 'body'> & {
   // Next.js stores body as buffer, while we store it as base64 encoded string
   body: string
   // Next.js doesn't produce cache-control tag we use to generate cdn cache control
   // so store needed values as part of cached response data
-  revalidate: Parameters<IncrementalCache['set']>[2]['revalidate']
+  revalidate?: Parameters<CacheHandler['set']>[2]['revalidate']
 }
 
-export type NetlifyCachedAppPageValue = Omit<IncrementalCachedAppPageValue, 'rscData'> & {
+/**
+ * Shape of the cache value that is returned from CacheHandler.get or passed to CacheHandler.set
+ */
+type IncrementalCachedAppPageValueForMultipleVersions = Omit<
+  IncrementalCachedAppPageValue,
+  'kind'
+> & {
+  kind: 'APP_PAGE'
+}
+
+/**
+ * Used for storing in blobs and reading from blobs
+ */
+export type NetlifyCachedAppPageValue = Omit<
+  IncrementalCachedAppPageValueForMultipleVersions,
+  'rscData'
+> & {
   // Next.js stores rscData as buffer, while we store it as base64 encoded string
   rscData: string | undefined
-  revalidate?: Parameters<IncrementalCache['set']>[2]['revalidate']
+  revalidate?: Parameters<CacheHandler['set']>[2]['revalidate']
 }
 
-type CachedPageValue = Extract<IncrementalCacheValue, { kind: 'PAGE' }>
-
-export type NetlifyCachedPageValue = CachedPageValue & {
-  revalidate?: Parameters<IncrementalCache['set']>[2]['revalidate']
+/**
+ * Shape of the cache value that is returned from CacheHandler.get or passed to CacheHandler.set
+ */
+type IncrementalCachedPageValueForMultipleVersions = Omit<IncrementalCachedPageValue, 'kind'> & {
+  kind: 'PAGE' | 'PAGES'
 }
 
-export type CachedFetchValue = Extract<IncrementalCacheValue, { kind: 'FETCH' }>
+/**
+ * Used for storing in blobs and reading from blobs
+ */
+export type NetlifyCachedPageValue = IncrementalCachedPageValueForMultipleVersions & {
+  revalidate?: Parameters<CacheHandler['set']>[2]['revalidate']
+}
 
-export type NetlifyIncrementalCacheValue =
-  | Exclude<
-      IncrementalCacheValue,
-      CachedRouteValue | CachedPageValue | IncrementalCachedAppPageValue
-    >
-  | NetlifyCachedRouteValue
-  | NetlifyCachedPageValue
-  | NetlifyCachedAppPageValue
+export type CachedFetchValueForMultipleVersions = Omit<CachedFetchValue, 'kind'> & {
+  kind: 'FETCH'
+}
```

Comment:
This one is really just to satisfy TS. Next now uses `const enum` with string values and you can't assign string literal to it (even tho the value matches).

For other similar cases like this - i.e. `kind: 'PAGE' | 'PAGES'` or `kind: 'ROUTE' | 'APP_ROUTE'` - Next actually changed the string value from `PAGE` to `PAGES` and `ROUTE` to `APP_ROUTE` and we want to support both.

For our `CacheHandler` implementation we do receive appropriate kind used by Next.js in `.set` variant as it's Next.js that produce cache value to store.

For `.get` case we check what we stored in handle things appropriately for kinds before and after `kind` change and we don't need to make any decision which should be used there.

We do have to figure out which `kind` to actually use when seeding blobs from prerendered content, so that's the only case that need to decide which `kind` to use (and it's done based on Next.js version)